Source/Core/DiscIO/FileSystemGCWii.cpp

```diff
@@ -19,83 +21,309 @@
+	return FindFileInfo(path, m_root);
 }
 
-const std::string CFileSystemGCWii::GetFileName(u64 _Address)
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(const std::string& path, const IFileInfo& file_info) const
 {
-	if (!m_Initialized)
-		InitFileSystem();
+	// Given a path like "directory1/directory2/fileA.bin", this method will
+	// find directory1 and then call itself to search for "directory2/fileA.bin".
+
+	// If there's nothing left to search for, return the current entry
+	if (path.empty())
+		return std::unique_ptr<IFileInfo>(new CFileInfoGCWii((const CFileInfoGCWii&)file_info));
 
-	for (auto& fileInfo : m_FileInfoVector)
+	// It's only possible to search in directories. Searching in a file is an error
+	if (!file_info.IsDirectory())
+		return nullptr;
+
+	std::string searching_for;
+	std::string rest_of_path;
+	size_t first_dir_separator = path.find('/');
+	if (first_dir_separator != std::string::npos)
+	{
+		searching_for = path.substr(0, first_dir_separator);
+		rest_of_path = path.substr(first_dir_separator + 1);
+	}
+	else
+	{
+		searching_for = path;
+		rest_of_path = "";
+	}
+
+	for (const IFileInfo& child : file_info)
 	{
-		if ((fileInfo.m_Offset <= _Address) &&
-		    ((fileInfo.m_Offset + fileInfo.m_FileSize) > _Address))
+		if (child.GetName() == searching_for)
 		{
-			return fileInfo.m_FullPath;
+			// A match is found. The rest of the path is passed on to finish the search
+			std::unique_ptr<IFileInfo> result = FindFileInfo(rest_of_path, child);
+			if (result)
+				return result;
 		}
 	}
 
-	return "";
+	return nullptr;
+}
+
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(u64 disc_offset) const
+{
+	if (!IsValid())
+		return nullptr;
+
+	if (m_offset_file_info_cache.empty())
+		BuildOffsetFileInfoCache(m_root);
+
+	// Get the first file that ends after disc_offset
+	auto it = m_offset_file_info_cache.upper_bound(disc_offset);
+	if (it == m_offset_file_info_cache.end())
+		return nullptr;
+	std::unique_ptr<IFileInfo> result = std::unique_ptr<IFileInfo>(new CFileInfoGCWii(m_root, it->second));
+
+	// If the file's start isn't after disc_offset, success
+	if (result->GetOffset() <= disc_offset)
+		return result;
+
+	return nullptr;
 }
 
-u64 CFileSystemGCWii::ReadFile(const std::string& _rFullPath, u8* _pBuffer, u64 _MaxBufferSize, u64 _OffsetInFile)
+void CFileSystemGCWii::BuildOffsetFileInfoCache(const IFileInfo& directory) const
 {
-	if (!m_Initialized)
-		InitFileSystem();
+	for (const IFileInfo& file_info : directory)
+	{
+		if (file_info.IsDirectory())
+			BuildOffsetFileInfoCache(file_info);
+		else
+			m_offset_file_info_cache.emplace(file_info.GetOffset() + file_info.GetSize(),
+		                                     reinterpret_cast<const CFileInfoGCWii&>(file_info).m_index);
+	}
+}
 
-	const SFileInfo* pFileInfo = FindFileInfo(_rFullPath);
-	if (pFileInfo == nullptr)
+u64 CFileSystemGCWii::ReadFile(const IFileInfo* file_info, u8* _pBuffer, u64 _MaxBufferSize, u64 _OffsetInFile) const
```

Comment:
While at it, can you fix the arguments name here for coding style compliance? I know it's not really your code, but... :)